osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs

```diff
@@ -101,8 +104,14 @@ public void Update(Playfield playfield)
 
             if (requiresHold)
                 changeState(true);
-            else if (isDownState && time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY)
-                changeState(false);
+            else
+            {
+                bool hasCompleted = time >= lastHitTime - relax_leniency;
```

Comment:
Isn't this completely wrong? Shouldn't it be `time > lastHitTime + relax_leniency`, or instead checking the result status of that object? As far as I can tell this is going to mean if the user arrives at the last object later than the time you propose here, relax is not going to hit the object for them.

Reply:
There's still the isDownState check right?

Reply:
In which case I don't understand what this is doing. Needs inline commenting explaining.

Reply:
Sounds like these should be refactored to:
```csharp
bool fulfilledReleaseDelay = time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY;
bool reachedLastObject = time >= lastHitTime - relax_leniency;

if (isDownState && (fulfilledReleaseDelay || reachedLastObject))
    changeState(false);
```